test/core/iomgr/pollset_windows_starvation_test.cc

```diff
@@ -0,0 +1,119 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#if defined(GRPC_WINSOCK_SOCKET)
+
+#include <vector>
+
+#include <grpc/grpc.h>
+#include <grpc/support/time.h>
+
+#include "src/core/lib/gprpp/thd.h"
+#include "src/core/lib/iomgr/exec_ctx.h"
+#include "src/core/lib/iomgr/iocp_windows.h"
+#include "src/core/lib/iomgr/iomgr_internal.h"
+#include "src/core/lib/iomgr/pollset.h"
+#include "src/core/lib/iomgr/pollset_windows.h"
+#include "src/core/lib/surface/init.h"
+#include "test/core/util/test_config.h"
+
+struct ThreadParams {
+  gpr_cv cv;
+  gpr_mu mu;
+  int complete;
+};
+
+int main(int argc, char** argv) {
+  grpc_init();
+
+  // Create three threads that all start queueing for work.
+  //
+  // The first one becomes the active poller for work and the two other
+  // threads go into the poller queue.
+  //
+  // When work arrives, the first one notifies the next active poller,
+  // this wakes the second thread - however all this does is return from
+  // the grpc_pollset_work function. It's up to that thread to figure
+  // out if it still wants to queue for more work or if it should kick
+  // other pollers.
+  //
+  // Previously that kick only affected pollers in the same pollset, thus
+  // leaving the third thread stuck in the poller queue. Now the pollset-
+  // specific grpc_pollset_kick will also kick pollers from other pollsets
+  // if there are no pollers in the current pollset. This frees up the
+  // last thread and completes the test.
+  ThreadParams params = {};
+  gpr_cv_init(&params.cv);
+  gpr_mu_init(&params.mu);
+  std::vector<grpc_core::Thread> threads;
+  for (int i = 0; i < 3; i++) {
+    grpc_core::Thread thd(
+        "Poller",
+        [](void* params) {
+          ThreadParams* tparams = static_cast<ThreadParams*>(params);
+          grpc_core::ExecCtx exec_ctx;
+
+          gpr_mu* mu;
+          grpc_pollset pollset = {};
+          grpc_pollset_init(&pollset, &mu);
+
+          gpr_mu_lock(mu);
+
+          // Queue for work and once we're done, make sure to kick the remaining
+          // threads.
+          grpc_millis deadline = grpc_timespec_to_millis_round_up(
+              grpc_timeout_seconds_to_deadline(5));
+          grpc_error* error;
+          error = grpc_pollset_work(&pollset, NULL, deadline);
+          error = grpc_pollset_kick(&pollset, NULL);
+
+          gpr_mu_unlock(mu);
+
+          {
+            gpr_mu_lock(&tparams->mu);
+            tparams->complete++;
+            gpr_cv_signal(&tparams->cv);
+            gpr_mu_unlock(&tparams->mu);
+          }
+        },
+        &params);
+    thd.Start();
+    threads.push_back(std::move(thd));
+  }
+
+  // Wait for the threads to start working and then kick one of them.
+  gpr_sleep_until(grpc_timeout_milliseconds_to_deadline(10));
```

Comment:
To be honest, I do not think this is a reliable way to wait for all the three threads to have been blocked in grpc_pollset_work. If any of the thread has not started grpc_pollset_work at this point, it will be stuck there later for 5 seconds, which will make this test flaky.

However, I do not have a good answer to completely eliminate this flakiness. A few suggestions:
1. I think two threads should be enough for this test. This will reduce the contention on the mutex at Ln 74;
2. Move the deadline calculation out of the critical section. This will ensure that each thread can finish its critical section sooner.
3. Consider extending this sleep time, which will give us a higher chance that both threads have entered grpc_pollset_work.

@jtattermusch Do you have any other suggestions?

Reply:
Two threads isn't guaranteed to fail without the fix.

|Time|event|T1|T2|T3|
|-|-|-|-|-|
|0|Start|Queued|Wait|Wait|
|1|External Kick|Run|Wait|Wait|
|2|Internal Kick|Finish|Run|Wait|
|3|T1/T2 Kick|Finish|Finish|Wait|

Without the third thread everything would work even without the global kick. Once the queued thread exits the poll-loop, it the pollset internals will kick the next available thread (even across pollsets). This will wake the second thread. It's the "this thread is done, kick something else" kick that fails to wake up the third thread, allowing the first two to finish while the third is still waiting.

--------

I'm adding logic into this to synchronize the test runner thread with the `grpc_pollset_work` call by passing the thread-local mutex back from the thread, waiting for the thread to lock these these and then waiting for them to be released by the `grpc_pollset_work` when the threads have been queued.

Reply:
Also replacing the deadlines with infinite timeouts. The 1-5 second deadlines were there mostly to ensure the failure case would be detected quicker, but given the risk of flakiness, I guess we can rely on the 5-minute "test is hanging, let's terminate it" logic provided by the test runner.

The success case is still 0.1sec according to the test runner, it's on the failure case that gets delayed.